main/main.cpp

```diff
@@ -12,230 +12,201 @@
 
+	const adcOutput adcReading = adc.readADC();
+
+	if (!deviceConnected)
+		return;
+
+	uint8_t buffer[STATUS_SIZE + NUM_CHAN * SAMPLE_SIZE + CRC_FLAG_SIZE];
+	memcpy(buffer, &adcReading.status, STATUS_SIZE);
+
+	const int32_t channelValue[NUM_CHAN]{adcReading.ch0, adcReading.ch1, adcReading.ch2,
+	                                     adcReading.ch3};
+	for (int i = 0; i < NUM_CHAN; ++i) {
+		memcpy(&buffer[STATUS_SIZE + (i * SAMPLE_SIZE)], &channelValue[i], SAMPLE_SIZE);
+	}
 	// Pack CRC match as a single byte
-	buffer[14] = (uint8_t)(adcReading.crc_match ? 1 : 0);
+	buffer[STATUS_SIZE + NUM_CHAN * SAMPLE_SIZE] = adcReading.crc_match;
 
 	// Only place ADC reading if the the BLE device is connected.
 	// TODO - this probably should be handled in a different way.
-	if (deviceConnected) {
-		xStreamBufferSend(xStreamBuffer, buffer, sizeof(buffer), 0);
-
-		// When the buffer is sufficiently large, time to send data.
-		// BLE allows to extend data packet from 27 to 251 bytes
-		// Schedule a task when buffer is closer to the upper limit
-		// TODO figure out the best number or a way to know when to schedule
-		if (xStreamBufferBytesAvailable(xStreamBuffer) > 200) {
-			xTaskNotifyGive(xHandleNotifyBLE);
-		}
+	xStreamBufferSend(adcStreamBufferHandle, buffer, sizeof(buffer), 0);
+
+	// When the buffer is sufficiently large, time to send data.
+	// BLE allows to extend data packet from 27 to 251 bytes
+	// Schedule a task when buffer is closer to the upper limit
+	// TODO figure out the best number or a way to know when to schedule
+	if (xStreamBufferBytesAvailable(adcStreamBufferHandle) > BLE_PUBL_DATA_TRIGGER) {
+		xTaskNotifyGive(bleAdcFeedPublisherTaskHandle);
 	}
 }
 
 // Task that handles calling the read adc function and placing the values in the buffer.
-void task_adc_read_and_buffer(void *parameter) {
+static void taskAdcReadAndBuffer(void *) {
 
 	while (true) {
-		// Wait until ISR notifies this task, or time out. ISR can notify multiple times,
-		// but this will reset the counter to zero each time.
-		// How many times did the ISR notify this task
-		uint32_t ulNotifiedValue = ulTaskNotifyTake(pdTRUE, portMAX_DELAY);
-		if (ulNotifiedValue > 0) {
-			adc_read_and_buffer();
+		// Wait until ISR notifies this task. Normally numNotifications == 1,
+		// numNotifications > 1 in case we cannot keep up with adc and have some data lost.
+		uint32_t numNotifications = ulTaskNotifyTake(pdTRUE, portMAX_DELAY);
+		if (numNotifications > 0) [[likely]] {
+			adcReadAndBuffer();
 		}
 	}
 	vTaskDelete(NULL);
 }
 
 // Read the adc buffer and update the BLE characteristic
-void notify_ble_adc_buffer() {
-	// TODO verify that 251 is the actual max that can be sent and not
-	// 251 minus some header.
+static void blePublishAdcBuffer() {
 	// TODO figure if should check deviceConnected?
-	uint8_t batch[251];
-
-	size_t bytesRead = xStreamBufferReceive(xStreamBuffer, batch, sizeof(batch), 0);
-
+	uint8_t batch[BLE_PUBL_DATA_DLE];
```

Comment:
`BLE_PUBL_DATA_DLE` should actually be `BLE_PUBL_DATA_ATT_PAYLOAD`, since one packet can't send more than the payload size.
Also, add a comment why we are doing this:
```
// To maximize throughput, we want to send the most amount of data in a packet without having be split into multiple packets over the air.
```

Reply:
Also, we should probably change this from the maximum payload size to a multiple of the ADC reading size.
But this can be done in a different PR. I've created #15 for it.

Reply:
Done.